[#74] 찜 목록 조회 기능 구현

hanwix2 opened the pull request:
#74 

## 개요
- 찜 목록 조회
    - 현재 로그인 된 user의 찜 목록
    - 전체 & 공연타입별 조회 가능
    - 커서(Cursor) 기반 페이징 처리
- _일반 회원 기능_

## Progress
- [x] 기본 찜 목록 조회 
- [x] 커서 방식 페이징 적용
- [x] 공연 타입별 목록 조회
- [x] 테스트 코드

---

1. **찜 목록 조회**
    - 기본적인 서비스 메커니즘은 [#84]**공연 리스트 조회** 기능과 유사
    - pick 테이블의 userId에 매칭되는 공연(performance)를 모두 조회

2.  **커서 기반 페이징 처리**
    - 공연 조회 페이징 설정 관련 `PerformancePagingCriteria` 클래스를 기반으로 직전에 마지막으로 조회된 공연(**lastPerfId**) 이후의 지정된 수(amount) 만큼 찜 목록 반환
    - **lastPerfId**는 RequestParam으로 받고, 만약 Null 일 시 찜 목록의 처음부터 지정된 수 만큼 반환

3. **공연 타입별 목록 조회**
    - 공연타입(**ShowType**)을 RequestParam으로 전달받아 해당 타입의 공연 리턴
    - **ShowType**이 Null 이라면 전체 찜 목록 리턴

4. ~~**캐시 적용**~~
    1. ~~**전체 찜 목록 조회(MAIN_PICK_LIST):**~~
        - ~~첫 찜 목록 페이지~~
        - ~~캐시 유지 시간: 30분~~
        - ~~찜 목록은 모든 공연 리스트를 조회하는 것 보다 데이터 수가 현저히 적음~~
        -  ~~저장된 찜의 공연 정보를 모두 확인하는 시간을 30분이라 예상. (리스트와 각 공연상세 정보 조회를 번갈아가며 확인)~~
    2. ~~**공연 타입 별 찜 목록 조회(TYPED_PICK_LIST):**~~
        - ~~공연 타입을 구분하여 조회하는 첫 페이지~~
        - ~~캐시 유지 시간: 20분~~
        - ~~공연 타입을 구분을 하여 조회를 하면 전체 조회보다도 더 데이터가 줄어들어 상대적으로 전체 찜 목록 조회보다 짧게 설정~~
    3. **etc**
        - ~~찜 목록 추가, 삭제 기능 수행 시 찜 조회 관련 모든 캐시 해제~~
        - ~~조회 중 첫번째 페이지에 대한 것만 캐시 적용~~
        - ~~lastPerfId 값에 따라 페이지를 구분하여 목록을 조회하는 경우엔 적은 데이터량으로 인한 사용 빈도가 낮다고 생각하여 캐시 적용 제외~~
        - **정적인 데이터가 아닌 회원마다 다른 결과를 응답하기 때문에 캐시 적용 불필요**

hanwix2 committed:
[#74] 찜 목록 조회 기능 구현

- 현재 로그인 된 User의 찜 목록 조회
- pick 테이블의 userId에 매칭되는 공연(performance)를 모두 조회
- 기본적인 서비스 메커니즘은 [#84]공연 리스트 조회 기능과 유사

hanwix2 committed:
[#74] 찜 목록 조회 시 커서 기반 페이징 처리

- PerformancePagingCriteria 클래스를 기반으로 직전에 마지막으로 조회된 공연(lastPerfId) 이후의 지정된 수(amount) 만큼 찜 목록 반환
- lastPerfId는 RequestParam으로 받고, 만약 Null 값일 시 찜 목록의 처음부터 지정된 수 만큼 반환

hanwix2 committed:
[#74] 공연 타입별 찜 목록 조회 기능 구현

- 공연타입(ShowType)을 RequestParam으로 전달받아 해당 타입의 공연 리턴

f-lab-dev commented on src/main/resources/mybatis/mappers/PerformanceMapper.xml:
```diff
@@ -170,4 +170,27 @@
     </select>
 
+    <select id="getPickedPerformances" resultMap="performanceResponse">
+        SELECT
```
이 부분도 실행계획을 뽑아보면 좋겠네요~

hanwix2 committed:
[#74] PickService 테스트 코드 작성 & 트랜잭션 적용

hanwix2 committed:
[#74] 찜 목록 조회 캐시 적용

1. 전체 찜 목록 조회(MAIN_PICK_LIST):
    - 첫 찜 목록 페이지
    - 캐시 유지 시간: 30분
    - 찜 목록은 모든 공연 리스트를 조회하는 것 보다 데이터 수가 현저히 적고,
      저장된 찜의 공연 정보를 모두 확인하는 시간을 30분이라 예상
2. 공연 타입 별 찜 목록 조회(TYPED_PICK_LIST)
    - 공연 타입을 구분하여 조회하는 첫 페이지
    - 캐시 유지 시간: 20분
    - 공연 타입을 구분을 하여 조회를 하면 전체 조회보다도 더 데이터가
      줄어들어 상대적으로 전체 찜 목록 조회보다 짧게 설정
3. etc
    - 찜 목록 추가, 삭제 기능 수행 시 찜 조회 관련 모든 캐시 해제
    - 조회 중 첫번째 페이지에 대한 것만 캐시 적용
    - lastPerfId 값에 따라 페이지를 구분하여 목록을 조회하는 경우엔
      적은 데이터량으로 인한 사용 빈도가 낮다고 생각하여 캐시 적용 제외

hanwix2 committed:
[#74] 찜 목록 캐시 미적용

- 정적인 데이터가 아닌 회원마다 다른 결과를 응답하기 때문에 캐시 적용 불필요

f-lab-dev reviewed (approved):
찜 목록은 사용자 패턴을 봐야겠지만 반복해서 조회하는 패턴이 많은 경우 회원마다 다른 결과를 응답하더라도 약간의 캐싱을 넣는건 괜찮습니다~
